File: AliyunLogProducer/AliyunLogProducer/LogProducerClient.m
//
//  LogProducerClient.m
//  AliyunLogProducer
//
//  Created by lichao on 2020/9/27.
//  Copyright © 2020 lichao. All rights reserved.
//

#import <Foundation/Foundation.h>
#import "LogProducerClient.h"
#import "LogProducerConfig.h"
#import "Log.h"
#import "TimeUtils.h"

#if __has_include("LogProducerClient+Bricks.h")
#import "LogProducerClient+Bricks.h"
#import "TCData.h"
#endif


@interface LogProducerClient ()

@end

@implementation LogProducerClient

- (id) initWithLogProducerConfig:(LogProducerConfig *)logProducerConfig
{
    return [self initWithLogProducerConfig:logProducerConfig callback:nil];
}

- (id) initWithLogProducerConfig:(LogProducerConfig *)logProducerConfig callback:(on_log_producer_send_done_function)callback
{
    if (self = [super init])
    {
        self->producer = create_log_producer(logProducerConfig->config, *callback, nil);
        self->client = get_log_producer_client(self->producer, nil);
        
        NSString *endpoint = [logProducerConfig getEndpoint];
        NSString *project = [logProducerConfig getProject];
        if ([endpoint length] != 0 && [project length] != 0) {
            [TimeUtils startUpdateServerTime:endpoint project:project];
        }
        enable = YES;
    }

    return self;
}

- (void)DestroyLogProducer
{
    if (!enable) {
        return;
    }
    enable = NO;
    destroy_log_producer(self->producer);
}

- (LogProducerResult)AddLog:(Log *) log
{
    return [self AddLog:log flush:0];
}

- (LogProducerResult)AddLog:(Log *) log flush:(int) flush
{
    if (!enable || self->client == NULL || log == nil) {
        return LogProducerInvalid;
    }
    NSMutableDictionary *logContents = log->content;
    
#if __has_include("LogProducerClient+Bricks.h")
    if (self->_enableTrack) {
        TCData *data = [TCData createDefault];
        NSDictionary *fields = [data toDictionary] ;
        for (id key in fields) {
            [logContents setObject:[fields valueForKey:key] forKey:key];
        }
    } else {
        if(self ->addLogInterceptor) {
            addLogInterceptor(log);
        }
    }
#else
    if(self ->addLogInterceptor) {
        addLogInterceptor(log);
    }
#endif
    
    int pairCount = (int)[logContents count];
        
    char **keyArray = (char **)malloc(sizeof(char *)*(pairCount));
    char **valueArray = (char **)malloc(sizeof(char *)*(pairCount));
    
    int32_t *keyCountArray = (int32_t*)malloc(sizeof(int32_t)*(pairCount));
    int32_t *valueCountArray = (int32_t*)malloc(sizeof(int32_t)*(pairCount));
    
    
    int ids = 0;
    for (NSString *key in logContents) {
        NSString *value = logContents[key];

        char* keyChar=[self convertToChar:key];
        char* valueChar=[self convertToChar:value];

        keyArray[ids] = keyChar;
        valueArray[ids] = valueChar;
        keyCountArray[ids] = (int32_t)strlen(keyChar);
        valueCountArray[ids] = (int32_t)strlen(valueChar);
        
        ids = ids + 1;
    }
    log_producer_result res = log_producer_client_add_log_with_len_time_int32(self->client, log->logTime, pairCount, keyArray, keyCountArray, valueArray, valueCountArray, flush);
    
    for(int i=0;i<pairCount;i++) {
        free(keyArray[i]);
        free(valueArray[i]);
    }
    free(keyArray);
    free(valueArray);
    free(keyCountArray);
    free(valueCountArray);
    return res;
}

-(char*)convertToChar:(NSString*)strtemp
{
    NSUInteger len = [strtemp lengthOfBytesUsingEncoding:NSUTF8StringEncoding] + 1;
    if (len > 1000000) return strdup([strtemp UTF8String]);
    char cStr [len];
    [strtemp getCString:cStr maxLength:len encoding:NSUTF8StringEncoding];
    return strdup(cStr);
}

- (void) setAddLogInterceptor: (AddLogInterceptor *) addLogInterceptor {
    self -> addLogInterceptor = *addLogInterceptor;
}

@end

File: AliyunLogProducer/AliyunLogProducer/LogProducerClient.swift
//
//  LogProducerClient.swift
//  AliyunLogProducer
//
//  Created by lichao on 2020/8/25.
//  Copyright © 2020 lichao. All rights reserved.
//

import Foundation

public var LogProducerCallbackQueue = Queue<LogProducerCallbackResult>(100)

open class LogProducerClient: NSObject {
    
    fileprivate var producer: OpaquePointer!
    fileprivate var client: UnsafeMutablePointer<log_producer_client>!

    @objc
    public init(logProducerConfig:LogProducerConfig){
        if logProducerConfig.IsGlobalCallbackEnable {
            let callback: on_log_producer_send_done_function = {config_name,result,log_bytes,compressed_bytes,req_id,error_message,raw_buffer,user_param in
                let logProducerResult = LogProducerResult(result)
                let reqId = req_id == nil ? "":String(cString: req_id!)
                let errorMessage = error_message == nil ? "" : String(cString: error_message!)
                let logProducerCallbackResult = LogProducerCallbackResult(logProducerResult, reqId, errorMessage, log_bytes, compressed_bytes)
                LogProducerCallbackQueue.enqueue(logProducerCallbackResult)
            }
            producer = create_log_producer(logProducerConfig.logProducerConfig, callback, nil)
        } else {
            producer = create_log_producer(logProducerConfig.logProducerConfig, nil, nil)
        }
        client = get_log_producer_client(producer, nil);
    }
    
    @objc
    public init(logProducerConfig:LogProducerConfig, callback:@escaping on_log_producer_send_done_function){
        producer = create_log_producer(logProducerConfig.logProducerConfig, callback, nil)
        client = get_log_producer_client(producer, nil);
    }
    
    /**
     add log to producer.

     - parameter log: Log
    */
    @objc
    open func AddLog(_ log: Log) -> LogProducerResult? {
        return AddLog(log, flush:0)
    }
    
    /**
     add log to producer.
    
     - parameter log Log
     - parameter flush: Int32 if this log info need to send right, 1 mean flush and 0 means NO
    */
    @objc
    open func AddLog(_ log: Log, flush: Int32) -> LogProducerResult? {
        if client == nil {
            return LogProducerResult(LogProducerEnum.LOG_PRODUCER_INVALID.rawValue)
        }
        let logContents = log.content
        let pairCount = logContents.count
        
        var keyArray = [String]()
        var valueArray = [String]()
        
        var keyCountArray = [Int]()
        var valueCountArray = [Int]()
        
        for logContent in logContents {
            let keyCount = logContent.key.lengthOfBytes(using: String.Encoding.utf8);
            let valueCount = logContent.value.lengthOfBytes(using: String.Encoding.utf8);
            
            keyCountArray.append(keyCount)
            valueCountArray.append(valueCount)
            
            keyArray.append(logContent.key)
            valueArray.append(logContent.value)
        }
        var keyPointerArrayPointer = keyArray.map { strdup($0) }
        var valuePointerArrayPointer = valueArray.map { strdup($0) }
        
        let res = log_producer_client_add_log_with_len(client, Int32(pairCount), &keyPointerArrayPointer, &keyCountArray, &valuePointerArrayPointer, &valueCountArray, flush)
        
        keyPointerArrayPointer.forEach() { free($0) }
        valuePointerArrayPointer.forEach() { free($0) }
        
        return LogProducerResult(res)
    }
    
    @objc
    open func DestroyLogProducer() {
        destroy_log_producer(producer);
    }
    
    deinit {
        self.DestroyLogProducer()
    }
}
